core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1624,6 +1632,46 @@ private RexNode simplifyOrs(List<RexNode> terms, RexUnknownAs unknownAs) {
             continue;
           }
         }
+        break;
+      case NOT_EQUALS:
+        final Comparison notEqualsComparison = Comparison.of(term);
+        if (notEqualsComparison != null) {
+          // We are dealing with a X<>A term, check if we saw before another NOT_EQUALS involving X
+          final RexNode prevNotEquals = notEqualsComparisonMap.get(notEqualsComparison.ref);
+          if (prevNotEquals == null) {
+            // This is the first NOT_EQUALS involving X, put it in the map
+            notEqualsComparisonMap.put(notEqualsComparison.ref, term);
+          } else {
+            // There is already in the map another NOT_EQUALS involving X:
+            //   - if it is already an IS_NOT_NULL: it was already simplified, ignore this term
+            //   - if it is not an IS_NOT_NULL (i.e. it is a NOT_EQUALS): check comparison values
+            if (prevNotEquals.getKind() != SqlKind.IS_NOT_NULL) {
+              final Comparable comparable1 = notEqualsComparison.literal.getValue();
+              //noinspection ConstantConditions
+              final Comparable comparable2 = Comparison.of(prevNotEquals).literal.getValue();
+              //noinspection unchecked
+              if (comparable1.compareTo(comparable2) != 0) {
+                // X <> A OR X <> B => X IS NOT NULL OR NULL
+                final RexNode isNotNull =
+                    rexBuilder.makeCall(SqlStdOperatorTable.IS_NOT_NULL, notEqualsComparison.ref);
+                final RexNode constantNull = rexBuilder.constantNull();
```

Comment:
I missed this earlier; this is an untyped null literal - we should be using a typed null boolean
```
rexBuilder.makeNullLiteral(e.getType())
```

Reply:
beyond this; I don't think I've anything else to add - except that patch actually enhances simplification; so the summary could be more precise: like "Enhance simplification to handle  (x<>a or x<>b)" or something similar.

Reply:
Thanks for the feedback @kgyrtkirk .
I've corrected the null issue and changed the Jira & PR titles according to your suggestion.

Reply:
Let's not use `RexBuilder.constantNull()`, because it produces an un-typed NULL. It's better to create a NULL of type BOOLEAN; call `RexBuilder.makeNullLiteral(RelDataType)`.

I am reviewing https://issues.apache.org/jira/browse/CALCITE-2921 currently and there are plan differences depending on which of the above methods are used.

Reply:
Thanks for the comment, I have replaced constantNull with a NULL of type BOOLEAN